lib/screens/v2/disruption_diagram/builder.ex

```diff
@@ -0,0 +1,967 @@
+  # anything special with the ends. They don't need to be split out.
+  defp split_end_stops(builder) when builder.metadata.line == :blue, do: builder
+
+  defp split_end_stops(builder) do
+    in_diagram =
+      [
+        closure_ideal_indices(builder),
+        gap_ideal_indices(builder),
+        current_location_ideal_indices(builder)
+      ]
+      |> Enum.map(&MapSet.new/1)
+      |> Enum.reduce(&MapSet.union/2)
+
+    {leftmost_stop_index, rightmost_stop_index} = Enum.min_max(in_diagram)
+
+    # Example: If the first one we're keeping is at index 5,
+    # then it's the 6th element so we need to slice off the first 5.
+    left_slice_amount = leftmost_stop_index
+
+    last_index = Vector.size(builder.sequence) - 1
+    right_slice_amount = last_index - rightmost_stop_index
+
+    builder
+    |> split_end(:right_end, right_slice_amount)
+    |> split_end(:left_end, left_slice_amount)
+    |> recalculate_metadata()
+  end
+
+  defp split_end(builder, end_field, 0), do: %{builder | end_field => Vector.new()}
+
+  defp split_end(builder, :left_end, amount) do
+    {left_end, sequence} = Vector.split(builder.sequence, amount)
+
+    # (We expect recalculate_metadata to be invoked in the calling function, so don't do it here.)
+    %{builder | sequence: sequence, left_end: left_end}
+  end
+
+  defp split_end(builder, :right_end, amount) do
+    {sequence, right_end} = Vector.split(builder.sequence, -amount)
+
+    %{builder | sequence: sequence, right_end: right_end}
+  end
+
+  # Re-computes index fields (home_stop, first/last_disrupted_stop)
+  # in builder.metadata after builder.sequence is changed.
+  #
+  # This function must be called after any operation that changes builder.sequence.
+  defp recalculate_metadata(builder) do
+    # We're going to replace all of the indices, so throw out the old ones.
+    # That way, if we fail to set one of them (which shouldn't happen),
+    # the `struct!` call below will fail instead of continuing with missing data.
+    meta_without_indices =
+      builder.metadata
+      |> Map.from_struct()
+      |> Map.drop([:home_stop, :first_disrupted_stop, :last_disrupted_stop])
+
+    new_metadata =
+      builder.sequence
+      |> Vector.with_index()
+      |> Vector.foldl(meta_without_indices, fn
+        {%StopSlot{} = stop_data, i}, meta ->
+          new_entries =
+            [
+              stop_data.disrupted? and [first_disrupted_stop: i, last_disrupted_stop: i],
+              stop_data.home_stop? and [home_stop: i]
+            ]
+            |> Enum.filter(& &1)
+            |> List.flatten()
+            |> Map.new()
+
+          Map.merge(meta, new_entries, fn
+            # This lets us set first_disrupted_stop only once.
+            :first_disrupted_stop, prev_value, _new_value -> prev_value
+            # The other entries get updated whenever there's a new value.
+            _k, _prev_value, new_value -> new_value
+          end)
```

Comment:
I feel a bit lost here. I get a vague sense of what is happening and the expected output, but am a bit thrown by the new_entries and why `first_disrupted_stop` only gets set once?

Reply:
`Vector.foldl` is the same as `Enum.reduce`, if that helps. You start with an initial accumulator, `meta_without_indices` in this case, and "reduce" the list down to a single value.

At a high level this is re-analyzing `sequence` to find the new indices of important slots, after `sequence` has been changed in some way.

For each slot in `sequence`, moving from left to right, if it's a `StopSlot` we check if that stop is disrupted or the home stop.

If it's disrupted:
- we try setting `first_disrupted_stop` if it wasn't already set. If it was, then that means this isn't the first one we've found!
- we always set `last_disrupted_stop`, because we want to keep updating that value until we reach the actual last disrupted stop.

If it's the home stop then we set the `home_stop`. There's only one of those.

I wrote the code this way to have the update happen in a single pass of `sequence`, but I agree it's not the clearest. Would this approach be better? It shouldn't have much of a performance impact.
```ex
# This code would replace the
# `new_metadata = builder.sequence |> Vector.with_index() |> Vector.foldl(...) |> then(...)`
# expression.

indexed_sequence = Vector.with_index(builder.sequence)

home_stop =
  Aja.Enum.find_value(indexed_sequence, fn
    {%StopSlot{home_stop?: true}, i} -> i
    _ -> false
  end)

first_disrupted_stop =
  Aja.Enum.find_value(indexed_sequence, fn
    {%StopSlot{disrupted?: true}, i} -> i
    _ -> false
  end)

last_disrupted_stop =
  indexed_sequence
  |> Vector.reverse()
  |> Aja.Enum.find_value(fn
    {%StopSlot{disrupted?: true}, i} -> i
    _ -> false
  end)

new_metadata =
  meta_without_indices
  |> Map.merge(%{
    home_stop: home_stop,
    first_disrupted_stop: first_disrupted_stop,
    last_disrupted_stop: last_disrupted_stop
  })
  |> then(&struct!(Metadata, &1))
```

Reply:
I do think the first round is great for its conciseness, but I much much prefer this rewrite for its readability!! And if it doesn't impact performance, then I think this would be better in the long-run